README
Github-Fork-Parent

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following command:
        cpan Github-Fork-Parent

        or

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Github::Fork::Parent

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Github-Fork-Parent

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Github-Fork-Parent

    CPAN Ratings
        http://cpanratings.perl.org/d/Github-Fork-Parent

    Search CPAN
        http://search.cpan.org/dist/Github-Fork-Parent/


COPYRIGHT AND LICENCE

Copyright (C) 2009-2014 Alexandr Ciornii

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Github-Fork-Parent

0.22    2014-05-01
        Update URL in test

0.21    2013-01-27
        Made online test check api.github.com availability first

0.20    2012-11-11
        Move to GitHub API v3
        Small fix for test

0.11    2011-05-09
        Parse https links too
        Modify tests to mirror github changes

0.10    2010-01-12
        Add 'github_parent_author'
        Fix error in test

0.01    2009-11-08
        First version.



lib/Github/Fork/Parent.pm
package Github::Fork::Parent;

use 5.006;
use strict;
use warnings;

=head1 NAME

Github::Fork::Parent - Perl module to determine which repository stands in a root of GitHub forking hierarhy.

=head1 VERSION

Version 0.22

=cut

our $VERSION = '0.22';


=head1 SYNOPSIS

my $parent_url = github_parent('git://github.com/chorny/cgi-session.git');
#returns https://github.com/cromedome/cgi-session

=head1 FUNCTIONS

=head2 github_parent

Takes link to repository (git://, git@ or http://) and returns http link to root repository.

=head2 github_parent_author

Takes link to repository (git://, git@ or http://) and returns owner of root repository.

=cut

use JSON;
#use YAML::Tiny 1.40;
use LWP::UserAgent;

use Exporter 'import';
our @EXPORT = qw(github_parent github_parent_author);

sub get_repo_data {
  my ($author,$project)=@_;
  #my $url = "http://github.com/api/v2/yaml/repos/show/$author/$project/network";
  my $url = "https://api.github.com/repos/$author/$project";

  my $ua=LWP::UserAgent->new();
  $ua->timeout(50);
  my $response = $ua->get($url);
  if ($response->is_success) {
    my $yaml = $response->content();
    return $yaml;
  } else {
    if ($response->code eq '404') {
      return undef;
    } else {
      die "Could not GET data (".$response->status_line.")";
    }
  }
}

sub parse_github_links {
  my $link=shift;
  if ($link=~m#^(?:\Qgit://github.com/\E|git\@github\.com:|https?://github\.com/)([^/]+)/([^/.]+)(?:\.git)?$#) {
    return ($1,$2);
  } else {
    return (undef,undef);
  }
  
}

sub github_parent {
  my $link=shift;
  my ($author,$project)=parse_github_links($link);
  return $link unless $author;
  my $yaml_content=get_repo_data($author,$project);
  if ($yaml_content) {
    #my $yaml=YAML::Tiny->read_string($yaml_content) or die;
    my $yaml=decode_json($yaml_content);
    my $source_url=$yaml->{source}{html_url};
    die unless $source_url;
    return $source_url;
  } else {
    die "No content";
  }
}

sub github_parent_author {
  #my $link=shift;
  #my $link1=github_parent($link);
  #my ($author,$project)=parse_github_links($link1);
  #die "Cannot get author from '$link1'" unless $author;
  #return $author;
  my $link=shift;
  my ($author,$project)=parse_github_links($link);
  return $link unless $author;
  my $yaml_content=get_repo_data($author,$project);
  if ($yaml_content) {
    #my $yaml=YAML::Tiny->read_string($yaml_content) or die;
    my $yaml=decode_json($yaml_content);
    my $source=$yaml->{source}{owner}{login};
    die unless $source;
    return $source;
    die;
  } else {
    die "No content";
  }
}

=head1 AUTHOR

Alexandr Ciornii, C<< <alexchorny at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-github-fork-parent at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Github-Fork-Parent>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Github::Fork::Parent


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Github-Fork-Parent>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Github-Fork-Parent>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Github-Fork-Parent>

=item * Search CPAN

L<http://search.cpan.org/dist/Github-Fork-Parent/>

=back


=head1 SEE ALSO

Net::GitHub

=head1 ACKNOWLEDGEMENTS


=head1 COPYRIGHT & LICENSE

Copyright 2009-2014 Alexandr Ciornii.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of Github::Fork::Parent


Build.PL
use 5.006;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Github::Fork::Parent',
    license             => 'perl',
    dist_author         => q{Alexandr Ciornii <alexchorny@gmail.com>},
    dist_version_from   => 'lib/Github/Fork/Parent.pm',
    requires => {
        'Exporter' => 5.57,
        'LWP::UserAgent' => 0,
        'YAML::Tiny' => 1.40,
        'perl' => 5.006,
    },
    build_requires => {
        'Test::More' => 0,
        'LWP::Online' => 1.06,
    },
    meta_merge => {
        resources=> {
            repository => 'http://github.com/chorny/Github-Fork-Parent',
        },
        keywords => ['github','repository','fork'],
    },
    add_to_cleanup      => [ 'Github-Fork-Parent-*' ],
    create_makefile_pl => 'traditional',
    auto_configure_requires => 0,
);

$builder->create_build_script();
